File: src/Data_Structure/Stack/Problems/MinimumAddtoMakeParenthesesValid/MinimumAdd.cpp
#include <iostream>
#include <stack>

using namespace std;

class Solution {
public:
    int minAddToMakeValid(string S) {

        int res = 0, bal = 0;
        for(char c: S)
            if(c == '(')
                res ++, bal ++;
            else{
                if(bal)
                    res --, bal --;
                else
                    res ++, bal = 0;
            }
        return res;
    }
};


int main() {

    cout << Solution().minAddToMakeValid("()))((") << endl;

    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: src/Data_Structure/Stack/Problems/MinimumAddtoMakeParenthesesValid/MinimumAdd.java
package Data_Structure.Stack.Problems.MinimumAddtoMakeParenthesesValid;

public class MinimumAdd {
    public int minAddToMakeValid(String s) {
        int count = 0;
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
          if (c == '(') {
            stack.push(c);
          } else {
            if (stack.isEmpty()) {
              count++;
            } else {
              stack.pop();
            }
          }
        }
        return count + stack.size();
      }
}

// Time Complexity : O(n)
// Space Complexity : O(1)

